#### Name: Rajnita Leichombam
#### USC-ID : 1370432160
#### Github-Username: Rajnita

# HOMEWORK - 3

## TIME SERIES CLASSIFICATION - PART I : Feature Creation/Extraction

(a)
Download the AReM data from:
https://archive.ics.uci.edu/ml/datasets/
Activity+Recognition+system+based+on+Multisensor+data+fusion+\%28AReM\
%29
. The dataset contains 7 folders that represent seven types of activities. In
each folder, there are multiple files each of which represents an instant of a human
performing an activity.
1
Each file containis 6 time series collected from activities
of the same person, which are called avg
rss12, var
rss12, avg
rss13, var
rss13,
vg
rss23, and ar
rss23. There are 88 instances in the dataset, each of which con-
tains 6 time series and each time series has 480 consecutive values

<span style="color:#E1721D">

The dataset is downloaded and put under '../data' folder.
    </span>

In [9]:
##ALL IMPORTS
import pandas as pd
import numpy as np
import bootstrapped.bootstrap as bs
import bootstrapped.stats_functions as bs_stats

(b)
Keep datasets 1 and 2 in folders bending1 and bending 2, as well as datasets 1,
2, and 3 in other folders as test data and other datasets as train data.

### CREATION OF TRAINING DATA AND TESTING DATA

For cycling/dataset14.csv and cycling/dataset9.csv, I removed the unnecessary ',' manually

In [10]:
list_of_folders_to_read=["bending1","bending2", "cycling","lying", "sitting", "standing","walking"]
time_series_columns=['time','avg_rss12','var_rss12','avg_rss13','var_rss13','avg_rss23','var_rss23']

##just for self check
train_file_paths=[]
training_data_files=[]
test_data_files=[]

train_dfs = list()
test_dfs = list()

def read_csv(i, directory,type_of_data):
    filename="/dataset"+str(i)+".csv"
    filepath=directory+filename
    if(type_of_data=='trainData'):
        training_data_files.append(filepath);
    else:
        ##test data
        test_data_files.append(filepath);
    if filepath=='../data/AReM/bending2/dataset4.csv':
        ##delimiting using whitespace specifically for this csv file
        df=pd.read_csv(filepath,comment='#', delim_whitespace=True, header=None);
    else:
        df=pd.read_csv(filepath,comment='#', header=None);
        
    return df;


##Used later in the notebook for 2c(i) to maintain order while printing all 88 instances, instead of merging train n test, using allinstances variable
all_instances=[]

for folder in list_of_folders_to_read:
    directory="../data/AReM/"+folder;
    if(folder.startswith('bending')):
        ##Train Data from 3,4,5...
        ##Test Data - 1,2
        ## bending1 has csv files of the format dataset<number>.csv 
        ##where number ranges from 1 to 7 in bending1 
        ##and 1 to 6 in bending2
        for i in range(1,3):
            df=read_csv(i, directory,'testData');
            df.columns=time_series_columns;
            df['activity']=folder
            test_dfs.append(df)
            all_instances.append(df)
            
        for i in range(3,7):
            df=read_csv(i, directory,'trainData');
            df.columns=time_series_columns;
            df['activity']=folder
            train_dfs.append(df)
            all_instances.append(df)

        if(folder=='bending1'):
            ##adding the 7th csv file in bending1
            df=read_csv(7, directory,'trainData');
            df.columns=time_series_columns;
            df['activity']=folder
            train_dfs.append(df)
            all_instances.append(df)
          
    else:
        ##Train Data from 4,5...
        ##Test Data - 1,2,3
        for i in range(1,4):
            df=read_csv(i, directory,'testData');
            df.columns=time_series_columns;
            df['activity']=folder
            test_dfs.append(df)
            all_instances.append(df)
        
        ##Each of the activities has 15 total sets of data
        for i in range(4,16):
            df=read_csv(i, directory,'trainData');
            df.columns=time_series_columns;
            df['activity']=folder
            train_dfs.append(df)
            all_instances.append(df)

print("\n################## TRAINING DATA #####################")
train_data = pd.concat(train_dfs, axis=0, ignore_index=True)
train_data


################## TRAINING DATA #####################


time  avg_rss12  var_rss12  avg_rss13  var_rss13  avg_rss23  \
0           0      42.00       0.71      21.25       0.43      30.00   
1         250      41.50       0.50      20.25       1.48      31.25   
2         500      41.50       0.50      14.25       1.92      33.00   
3         750      40.75       0.83      15.75       0.43      33.00   
4        1000      40.00       0.71      20.00       2.74      32.75   
...       ...        ...        ...        ...        ...        ...   
33114  118750      34.50       6.18       9.00       3.56      12.67   
33115  119000      25.75       6.02      13.75       2.05      16.00   
33116  119250      31.50       3.35      10.25       5.12      16.25   
33117  119500      33.75       2.77      14.00       3.24      13.75   
33118  119750      37.00       1.41      18.25       3.70      11.00   

       var_rss23  activity  
0           0.00  bending1  
1           1.09  bending1  
2           0.00  bending1  
3           0.00  bending1  
4           0.43  bending1  
...          ...       ...  
33114       4.19   walking  
33115       1.58   walking  
33116       2.95   walking  
33117       0.43   walking  
33118       4.32   walking  

[33119 rows x 8 columns]

#### The above training data has 33119 rows, one of the files had only 479 rows where all others had 480 rows

In [11]:
print("\n################## TESTING DATA #####################")
test_data = pd.concat(test_dfs, axis=0, ignore_index=True)
test_data


################## TESTING DATA #####################


time  avg_rss12  var_rss12  avg_rss13  var_rss13  avg_rss23  \
0          0      39.25       0.43      22.75       0.43      33.75   
1        250      39.25       0.43      23.00       0.00      33.00   
2        500      39.25       0.43      23.25       0.43      33.00   
3        750      39.50       0.50      23.00       0.71      33.00   
4       1000      39.50       0.50      24.00       0.00      33.00   
...      ...        ...        ...        ...        ...        ...   
9115  118750      36.00       2.45      17.00       5.10      20.50   
9116  119000      34.33       1.89      15.00       2.45      17.00   
9117  119250      33.00       7.35      14.60       3.14      13.00   
9118  119500      31.67       1.25      11.00       6.16      19.25   
9119  119750      30.75      10.21      11.75       1.09      18.50   

      var_rss23  activity  
0          1.30  bending1  
1          0.00  bending1  
2          0.00  bending1  
3          0.00  bending1  
4          0.00  bending1  
...         ...       ...  
9115       0.87   walking  
9116       2.12   walking  
9117       5.70   walking  
9118       2.17   walking  
9119       3.20   walking  

[9120 rows x 8 columns]

In [12]:
print("Total TRAINING FILES or Number of Training Instances = ",len(training_data_files))
print("\n############### LIST OF TRAINING FILES #################\n")
print(*training_data_files, sep = "\n")


Total TRAINING FILES or Number of Training Instances =  69

############### LIST OF TRAINING FILES #################

../data/AReM/bending1/dataset3.csv
../data/AReM/bending1/dataset4.csv
../data/AReM/bending1/dataset5.csv
../data/AReM/bending1/dataset6.csv
../data/AReM/bending1/dataset7.csv
../data/AReM/bending2/dataset3.csv
../data/AReM/bending2/dataset4.csv
../data/AReM/bending2/dataset5.csv
../data/AReM/bending2/dataset6.csv
../data/AReM/cycling/dataset4.csv
../data/AReM/cycling/dataset5.csv
../data/AReM/cycling/dataset6.csv
../data/AReM/cycling/dataset7.csv
../data/AReM/cycling/dataset8.csv
../data/AReM/cycling/dataset9.csv
../data/AReM/cycling/dataset10.csv
../data/AReM/cycling/dataset11.csv
../data/AReM/cycling/dataset12.csv
../data/AReM/cycling/dataset13.csv
../data/AReM/cycling/dataset14.csv
../data/AReM/cycling/dataset15.csv
../data/AReM/lying/dataset4.csv
../data/AReM/lying/dataset5.csv
../data/AReM/lying/dataset6.csv
../data/AReM/lying/dataset7.csv
../data/AReM/lying/datase

In [13]:
print("Total TEST FILES or Number of Testing Instances:",len(test_data_files))
print("\n###############LIST OF TEST FILES###############\n")
print(*test_data_files, sep = "\n")


Total TEST FILES or Number of Testing Instances: 19

###############LIST OF TEST FILES###############

../data/AReM/bending1/dataset1.csv
../data/AReM/bending1/dataset2.csv
../data/AReM/bending2/dataset1.csv
../data/AReM/bending2/dataset2.csv
../data/AReM/cycling/dataset1.csv
../data/AReM/cycling/dataset2.csv
../data/AReM/cycling/dataset3.csv
../data/AReM/lying/dataset1.csv
../data/AReM/lying/dataset2.csv
../data/AReM/lying/dataset3.csv
../data/AReM/sitting/dataset1.csv
../data/AReM/sitting/dataset2.csv
../data/AReM/sitting/dataset3.csv
../data/AReM/standing/dataset1.csv
../data/AReM/standing/dataset2.csv
../data/AReM/standing/dataset3.csv
../data/AReM/walking/dataset1.csv
../data/AReM/walking/dataset2.csv
../data/AReM/walking/dataset3.csv


In [14]:
print("TOTAL INSTANCES = ",len(train_dfs)+len(test_dfs))

TOTAL INSTANCES =  88


(c)
Feature Extraction
Classification of time series usually needs extracting features from them. In this
problem, we focus on time-domain features. <br>
i.
Research what types of time-domain features are usually used in time series
classification and list them (examples are minimum, maximum, mean, etc).

### TIME DOMAIN FEATURES USED IN TIME SERIES CLASSIFICATION

Some of the Time domain features that are used in time series classification are -<br>
1) Mean <br>
2) Standard Deviation <br>
3) Median<br>
4) Variance<br>
5) Minimum<br>
6) Maximum<br>
7)Skewnesss: measure of assymmetry of a signal<br>
8)Quartiles ( first quartile,third quartile)<br>
9) Auto Correlation -  correlation between two observations at different points in a time series<br>
10) Cross Correlation - measure of association between different time series<br>
11) Slope Sign Change - count of number of times that slope of a signal sign change<br>
12)Interquartile range<br>
13)Kurtosis: measure of peakness of a probability distribution<br>
14)Mean absolute deviation - avg of the absolute deviations of data from the data's mean<br>
15) Zero Crossing Rate - rate at which a signal changes its sign from positive to negative or vice versa within a given window<br>
16) Mean Crossing Rate - rate at which a signal changes from above mean to below mean and vice versa
<br>

<br><br>Extract the time-domain features minimum, maximum, mean, median, stan-
dard deviation, first quartile, and third quartile for all of the 6 time series
in each instance. You are free to normalize/standardize features or use them
directly.


### EXTRACTION OF TIME DOMAIN FEATURES 

In [15]:
train_dfs[0]
# all_instances=train_dfs+test_dfs
print("TOTAL INSTANCES: ",len(all_instances))

column_names_for_calculation=['avg_rss12','var_rss12','avg_rss13','var_rss13','avg_rss23','var_rss23']
time_dmn_features_all_instances=[]
columns_in_time_domain_features=[]

##Setting the column names of the dataframe with time domain features
for i in range(1,7):
    columns_in_time_domain_features.append('MIN_'+str(i));
    columns_in_time_domain_features.append('MAX_'+str(i));
    columns_in_time_domain_features.append('MEAN_'+str(i));
    columns_in_time_domain_features.append('MEDIAN_'+str(i));
    columns_in_time_domain_features.append('STD_'+str(i));
    columns_in_time_domain_features.append('FIRST_QUARTILE_'+str(i));
    columns_in_time_domain_features.append('THIRD_QUARTILE'+str(i));
                   
for instance in all_instances:
    stats_current_instance=[]
    for column in column_names_for_calculation:
        ##The order of statistics is as the order of columns created above at line 22
        stats_current_instance.append(instance[column].min())
        stats_current_instance.append(instance[column].max())
        stats_current_instance.append(instance[column].mean())
        stats_current_instance.append(instance[column].median())
        stats_current_instance.append(instance[column].std())
        stats_current_instance.append(instance[column].quantile(0.25))
        stats_current_instance.append(instance[column].quantile(0.75))
    
    time_dmn_features_all_instances.append(stats_current_instance);

time_dmn_features_dataframe=pd.DataFrame(time_dmn_features_all_instances, columns=columns_in_time_domain_features);
print("Total number of columns:",time_dmn_features_dataframe.columns.size)

print("*********************Time domain features of all instances: *********************\n")
with pd.option_context('display.max_rows', None,
                       'display.max_columns', None
                       ):
    display(time_dmn_features_dataframe)

TOTAL INSTANCES:  88
Total number of columns: 42
*********************Time domain features of all instances: *********************



MIN_1  MAX_1     MEAN_1  MEDIAN_1     STD_1  FIRST_QUARTILE_1  \
0   37.25  45.00  40.624792    40.500  1.476967           39.2500   
1   38.00  45.67  42.812812    42.500  1.435550           42.0000   
2   35.00  47.40  43.954500    44.330  1.558835           43.0000   
3   33.00  47.75  42.179813    43.500  3.670666           39.1500   
4   33.00  45.75  41.678063    41.750  2.243490           41.3300   
5   37.00  48.00  43.454958    43.250  1.386098           42.5000   
6   36.25  48.00  43.969125    44.500  1.618364           43.3100   
7   12.75  51.00  24.562958    24.250  3.737514           23.1875   
8    0.00  42.75  27.464604    28.000  3.583582           25.5000   
9   21.00  50.00  32.586208    33.000  6.238143           26.1875   
10  27.50  33.00  29.881938    30.000  1.153837           29.0000   
11  19.00  45.50  30.938104    29.000  7.684146           26.7500   
12  25.00  47.50  31.058250    29.710  4.829794           27.5000   
13  24.25  45.00  37.177042    36.250  3.581301           34.5000   
14  28.75  44.75  37.561188    36.875  3.226507           35.2500   
15  22.00  44.67  37.058708    36.000  3.710180           34.5000   
16  19.00  44.00  36.228396    36.000  3.528617           34.0000   
17  26.50  44.33  36.687292    36.000  3.529404           34.2500   
18  25.33  45.00  37.114312    36.250  3.710385           34.5000   
19  26.75  44.75  36.863375    36.330  3.555787           34.5000   
20  26.25  44.25  36.957458    36.290  3.434863           34.5000   
21  27.75  44.67  37.144833    36.330  3.758904           34.0000   
22  27.00  45.00  36.819521    36.000  3.900459           33.7500   
23  27.00  44.33  36.541667    36.000  4.018922           33.2500   
24  18.50  44.25  35.752354    36.000  4.614802           33.0000   
25  19.00  43.75  35.879875    36.000  4.614878           33.0000   
26  23.33  43.50  36.244083    36.750  3.822016           33.4575   
27  24.25  45.00  37.177042    36.250  3.581301           34.5000   
28  23.50  30.00  27.716375    27.500  1.442253           27.0000   
29  24.75  48.33  44.182937    48.000  7.495615           48.0000   
30  48.00  48.25  48.004167    48.000  0.032038           48.0000   
31  34.00  51.00  42.706063    40.500  3.537476           40.2500   
32  39.00  41.00  39.667833    39.500  0.280158           39.5000   
33   0.00  40.67  39.506187    39.500  1.817498           39.5000   
34  39.00  40.00  39.433792    39.500  0.208558           39.3300   
35  39.00  40.00  39.347104    39.330  0.231405           39.2500   
36  39.00  56.25  47.325125    42.670  5.961280           42.0000   
37  23.50  30.00  27.716375    27.500  1.442253           27.0000   
38  48.00  48.25  48.004167    48.000  0.032038           48.0000   
39  39.00  41.00  39.667833    39.500  0.280158           39.5000   
40  39.00  40.00  39.433792    39.500  0.208558           39.3300   
41  39.00  56.25  47.325125    42.670  5.961280           42.0000   
42  23.50  30.00  27.716375    27.500  1.442253           27.0000   
43  33.25  48.00  42.363563    43.000  2.068247           42.0000   
44  37.00  50.75  45.917667    45.635  1.089027           45.0000   
45  44.50  46.75  45.239667    45.000  0.429915           45.0000   
46  42.50  46.00  44.753958    45.000  0.520309           44.7500   
47  25.50  45.75  36.560063    36.500  5.915081           30.0000   
48  27.00  38.67  32.905979    30.250  3.505776           30.0000   
49  27.50  50.75  38.888375    36.750  3.642901           36.2500   
50  42.00  45.00  42.821983    42.500  0.891818           42.0000   
51  41.75  46.50  43.190854    42.500  1.230345           42.0000   
52  35.25  48.50  40.219271    39.250  2.741217           37.7500   
53  28.50  48.25  43.884833    45.250  3.198894           42.0000   
54  39.50  45.00  42.111583    42.000  1.122245           41.5000   
55  39.67  44.75  42.282667    41.500  1.356149           41.5000   
56  40.00  44.67  42.360187    42.000  1.017372           41.5000   
57

iii.
Estimate the standard deviation of each of the time-domain features you
extracted from the data. Then, use Python’s bootstrapped or any other
method to build a 90% bootstrap confidence interval for the standard deviation
of each feature.


### BUILDING 90% BOOTSTRAP CONFIDENCE INTERVAL FOR THE STANDARD DEVIATION OF EACH FEATURE

In [16]:
confidence_intervals_all=[];
for column in columns_in_time_domain_features:
    current_feature_ci=[]
    bs_results= bs.bootstrap(np.array(time_dmn_features_dataframe[column]), alpha=0.10, stat_func=bs_stats.std)
    current_feature_ci.append(column);
    current_feature_ci.append(time_dmn_features_dataframe[column].std())
    current_feature_ci.append(bs_results.lower_bound)
    current_feature_ci.append(bs_results.upper_bound)
   
    confidence_intervals_all.append(current_feature_ci)
    
print("Number of features:",len(confidence_intervals_all))
ci_df=pd.DataFrame(confidence_intervals_all, columns=['Feature','Standard Deviation', 'Lower Bound CI','Upper Bound CI'])
ci_df


Number of features: 42


Feature  Standard Deviation  Lower Bound CI  Upper Bound CI
0              MIN_1            9.569975        8.300660       10.810246
1              MAX_1            4.394362        3.479297        5.428480
2             MEAN_1            5.335718        4.762605        5.941910
3           MEDIAN_1            5.440054        4.849359        6.057164
4              STD_1            1.772153        1.585693        1.963669
5   FIRST_QUARTILE_1            6.153590        5.631169        6.689624
6    THIRD_QUARTILE1            5.138925        4.399915        5.912566
7              MIN_2            0.000000        0.000000        0.000000
8              MAX_2            5.062729        4.702972        5.468948
9             MEAN_2            1.574164        1.432401        1.741196
10          MEDIAN_2            1.412244        1.270893        1.576640
11             STD_2            0.884105        0.822471        0.959225
12  FIRST_QUARTILE_2            0.946386        0.853025        1.059393
13   THIRD_QUARTILE2            2.125266        1.949096        2.349973
14             MIN_3            2.956462        2.788854        3.130001
15             MAX_3            4.875137        4.266351        5.547031
16            MEAN_3            4.008380        3.508266        4.578526
17          MEDIAN_3            4.036396        3.506239        4.618417
18             STD_3            0.946710        0.767376        1.124506
19  FIRST_QUARTILE_3            4.220658        3.709192        4.786280
20   THIRD_QUARTILE3            4.171628        3.629409        4.789768
21             MIN_4            0.000000        0.000000        0.000000
22             MAX_4            2.183625        1.998297        2.377832
23            MEAN_4            1.166114        1.103889        1.249452
24          MEDIAN_4            1.145586        1.083285        1.228399
25             STD_4            0.458242        0.427565        0.492173
26  FIRST_QUARTILE_4            0.843620        0.791255        0.907698
27   THIRD_QUARTILE4            1.552504        1.469209        1.663089
28             MIN_5            6.124001        4.718197        7.799386
29             MAX_5            5.741238        4.887813        6.688033
30            MEAN_5            5.675593        4.624842        6.881469
31          MEDIAN_5            5.813782        4.694216        7.060194
32             STD_5            1.024898        0.830345        1.229696
33  FIRST_QUARTILE_5            6.096465        4.953702        7.358912
34   THIRD_QUARTILE5            5.531720        4.502895        6.651220
35             MIN_6            0.045838        0.013125        0.091154
36             MAX_6            2.518921        2.265273        2.770221
37            MEAN_6            1.154812        1.088068        1.241043
38          MEDIAN_6            1.086474        1.019400        1.172179
39             STD_6            0.517617        0.487768        0.552469
40  FIRST_QUARTILE_6            0.758584        0.704987        0.823091
41   THIRD_QUARTILE6            1.523599        1.437838        1.634997

iv.
Use your judgement to select the three most important time-domain features
(one option may be min, mean, and max).

I would choose <b>Mean, Standard Deviation and Median</b> as the three most important time features.
<br>
Mean is one of the simple time domain features that considers all the values in the dataset and give an average.<br>
Standard deviation shows the measurement of variability.<br>
From the research papers that I referred (put in the REFERENCES section in the last cell), mean and standard deviation are two good features that shows the differentiation in the body postures and the intensity of the activites.<br>
At times, mean can give a skewed value and is sensitive to extreme values. <br>
For activity monitoring data set, another time feature I would consider is one of the percentile values i.e. ( min-0th percentile, max-100 percentile, median - 50 percentile, first quartile or third quartile). 
I considered  Median  as another time feature, a good measure of central tendency in case of presence of extreme values.


## ISLR 3.7.4

#### a) True Relationship is Linear, training RSS lower for linear or cubic model??

<span style="color:green">
I would expect that Cubic Regression model will have lower training RSS.
With more number of polynomial terms, we know that cubic regression model will have a closer fit. In other words, higher degrees of freedom (more polynomial terms) implies, better fit to the data. 
Here,With the same data, we are comparing linear and cubic. I believe cubic model will overfit to any non linearities that are present in the training data and hence lead to lower RSS.
    
</span>


#### b) True Relationship is Linear, testing RSS lower for linear or cubic model??

<span style="color:green">
Given that the true relationship between X and Y is linear, in case of test data, I would expect the linear regressor to perform better i.e. lesser test RSS. Since the true model is linear, the cubic model may have overfitted the training data, and with test data the cubic model will likely have more error i.e. more test RSS.
</span>

#### c) True Relationship is not Linear (no idea how far), training RSS lower for linear or cubic model??

<span style="color:green">
A polynomial regression model will lead to closer fit as a result of greater flexibility compared to linear fit. Similar to (a), CUBIC fit will have LESSER TRAINING RSS, regardless of the information that true relationship is linear or non linear , as we are dealing with training data here.

</span>

#### d) True Relationship is not Linear (no idea how far), test RSS lower for linear or cubic model??


<span style="color:green">
I would say, it depends on how non linear the true relationship is. <br>
If the relationship is non linear but closer to linear, then I would expect linear regression to have a LOWER TEST RSS. In this case, cubic regression model would overfit and hence higher RSS.
<br>
If the relationship is more of a cubic nature than linear, then cubic regression model will have lesser RSS. Linear Regression model can underfit in this case and hence,higher RSS value.

</span>



REFERENCES:

https://pypi.org/project/bootstrapped/
https://revistia.org/files/articles/ejis_v2_i3s_16/Cemil.pdf
https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4014024 https://www.researchgate.net/publication/298911566_Human_activity_recognition_using_multisensor_data_fusion_based_on_Reservoir_Computing 
https://www.sciencedirect.com/science/article/pii/S0031320310001950
https://ieeexplore-ieee-org.libproxy2.usc.edu/stamp/stamp.jsp?tp=&arnumber=7250435    https://www.researchgate.net/publication/258029665_Multisensor_Data_Fusion_for_Activity_Recognition_Based_on_Reservoir_Computing
 https://www.researchgate.net/figure/Time-domain-features_tbl1_306269759
                        